studio/LonaStudio/Plugins/RPCService.swift

```diff
@@ -0,0 +1,184 @@
+    case alert
+}
+
+// A completion handler for a synchronous RPC
+typealias RpcCallback = (RpcResult) -> Void
+
+class RPCService {
+    // RPC state
+    private var queue = DispatchQueue(label: "com.devinabbott.ComponentStudio.CoreConnection", attributes: [])
+    private var rpcIndex = 0
+    private var pending = Dictionary<Int, RpcCallback>()
+
+    var sendData: ((_ data: Data) -> Void)?
+
+    private func sendJson(_ json: Any) {
+        do {
+            let data = try JSONSerialization.data(withJSONObject: json, options: [])
+            sendData?(data)
+        } catch _ {
+            print("error serializing to json")
+        }
+    }
+
+    private func sendResult(id: Any, result: Any) {
+        let json = ["id": id, "result": result]
+        sendJson(json)
+    }
+
+    private func sendError(id: Any, error: Any) {
+        let json = ["id": id, "error": error]
+        sendJson(json)
+    }
+
+    func handleRaw(_ data: Data) {
+        do {
+            let json = try JSONSerialization.jsonObject(with: data, options: .allowFragments)
+            handleRpc(json)
+        } catch {
+            print("json error \(error.localizedDescription)")
+            print(String(data: data, encoding: String.Encoding.utf8)!)
+        }
+    }
+
+    /// handle a JSON RPC call. Determines whether it is a request, response or notification
+    /// and executes/responds accordingly
+    private func handleRpc(_ json: Any) {
+        guard let obj = json as? [String: AnyObject] else { fatalError("malformed json \(json)") }
+        if let index = obj["id"] as? Int {
+            if obj["result"] != nil || obj["error"] != nil {
+                var callback: RpcCallback?
+                queue.sync {
+                    callback = self.pending.removeValue(forKey: index)
+                }
+                if let result = obj["result"] {
+                    callback?(.ok(result))
+                } else if let errJson = obj["error"] as? [String: AnyObject],
+                    let err = RemoteError(json: errJson) {
+                    callback?(.error(err))
+                } else {
+                    print("failed to parse response \(obj)")
+                }
+            } else {
+                self.handleRequest(json: obj)
+            }
+        } else {
+            self.handleNotification(json: obj)
+        }
+    }
+
+    private func handleRequest(json: [String: AnyObject]) {
+        guard let id = json["id"] else {
+                assertionFailure("unknown json from core: \(json)")
+                return
+        }
+        guard let jsonMethod = json["method"] as? String,
+              let method = RPCRequestMethod(rawValue: jsonMethod) else {
+                sendError(id: id, error: [
+                    "code": -32601,
+                    "message": "Method not found"
+                ])
+                return
+        }
+
+        switch method {
+        case .workspacePath:
```

Comment:
Let's move our own code out of the Xi stuff so it's easier to update their code if needed